service/history/version_membership_cache.go

```diff
@@ -0,0 +1,80 @@
+package history
+
+import (
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/server/common/cache"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/worker_versioning"
+)
+
+// versionMembershipCache is a typed wrapper around cache.Cache.
+// As of this implementation, this cache is present on every history host and stores results of the matching service's CheckTaskQueueVersionMembership RPC.
+// This cache emits cache hit/miss metrics.
+type versionMembershipCache struct {
+	cache   cache.Cache
+	metrics metrics.Handler
+}
+
+type versionMembershipCacheKey struct {
+	namespaceID    string
+	taskQueue      string
+	taskQueueType  enumspb.TaskQueueType
+	deploymentName string
+	buildID        string
+}
+
+func newVersionMembershipCache(c cache.Cache, metricsHandler metrics.Handler) worker_versioning.VersionMembershipCache {
+	h := metricsHandler.WithTags(metrics.CacheTypeTag("version_membership"))
```

Comment:
I think `"version_membership"` should be a constant in the `metrics` package, as in https://github.com/temporalio/temporal/blob/91893f10641bae49f8181235c5e93f61a3c826a8/service/history/events/cache.go#L87